File: backjoon/1463 1로 만들기.cc
// 다이나믹 프로그래밍, 배열로 메모리제이션
#include <iostream>
#include <stdio.h>
#include <algorithm>

using namespace std;

int memory[1000001] = { 0, };

int main()
{
    int N;
    scanf("%d", &N);
    memory[1] = 0;
    for (int i = 2; i <= N; i++)
    {
        int value = memory[i - 1];
        if (i % 2 == 0)
        {
            value = min(value, memory[i / 2]);
        }
        if (i % 3 == 0)
        {
            value = min(value, memory[i / 3]);
        }
        memory[i] = value + 1;
    }
    printf("%d\n", memory[N]);
    return (0);
}

File: backjoon/1463 1로 만들기.cs
using System;
using static System.Math;
using System.Collections;
namespace csharp
{
    class Program
    {
        static void Main(string[] args)
        {
            int[] memory = new int[1000001];
            int N = int.Parse(Console.ReadLine());
            Array.Fill<int>(memory, 0);
            memory[1] = 0;
            for (int i = 2; i <= N; i++)
            {
                int value = memory[i - 1];
                if (i % 2 == 0)
                {
                    value = Min(value, memory[i / 2]);
                }
                if (i % 3 == 0)
                {
                    value = Min(value, memory[i / 3]);
                }
                memory[i] = value + 1;
            }
            Console.WriteLine(memory[N]);
        }
    }
}
